crates/key-derivation/src/operator.rs

```diff
@@ -0,0 +1,482 @@
+
+        Ok(Self {
+            master: *master,
+            base: base_xpriv,
+            message: message_xpriv,
+            wallet: wallet_xpriv,
+        })
+    }
+
+    /// Operator's master [`Xpriv`].
+    pub fn master_xpriv(&self) -> &Xpriv {
+        &self.master
+    }
+
+    /// Operator's base [`Xpriv`].
+    ///
+    /// # Notes
+    ///
+    /// This is the [`Xpriv`] that is generated from only hardened paths from the master [`Xpriv`].
+    pub fn base_xpriv(&self) -> &Xpriv {
+        &self.base
+    }
+
+    /// Operator's wallet transaction signing [`Xpriv`].
+    pub fn wallet_xpriv(&self) -> &Xpriv {
+        &self.wallet
+    }
+
+    /// Operator's message signing [`Xpriv`].
+    pub fn message_xpriv(&self) -> &Xpriv {
+        &self.message
+    }
+
+    /// Operator's master [`Xpub`].
+    pub fn master_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.master)
+    }
+
+    /// Operator's base [`Xpub`]
+    ///
+    /// Infallible conversion from [`Xpriv`] to [`Xpub`] according to
+    /// [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+    ///
+    /// # Notes
+    ///
+    /// This is the [`Xpub`] that is generated from only hardened paths from the master [`Xpriv`].
+    pub fn base_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.base)
+    }
+
+    /// Operator's message signing [`Xpub`].
+    ///
+    /// Infallible according to
+    /// [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+    pub fn message_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.message)
+    }
+
+    /// Operator's wallet transaction signing [`Xpub`].
+    ///
+    /// Infallible conversion from [`Xpriv`] to [`Xpub`] according to
+    /// [BIP-32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki)
+    pub fn wallet_xpub(&self) -> Xpub {
+        Xpub::from_priv(SECP256K1, &self.wallet)
+    }
+}
+
+// Manual Drop implementation to zeroize keys on drop.
+impl Drop for OperatorKeys {
+    fn drop(&mut self) {
+        #[cfg(feature = "zeroize")]
+        self.zeroize();
+    }
+}
+
+#[cfg(feature = "zeroize")]
+impl Zeroize for OperatorKeys {
+    #[inline]
+    fn zeroize(&mut self) {
+        let Self {
+            master,
+            base,
+            message,
+            wallet,
+        } = self;
+
+        // # Security note
+        //
+        // Going over all possible "zeroizable" fields.
+        // What we cannot zeroize is only:
+        //
+        // - Network: enum
+        //
+        // These are fine to leave as they are since they are public parameters,
+        // and not secret values.
+        //
+        // NOTE: `Xpriv.private_key` (`SecretKey`) `non_secure_erase` writes `1`s to the memory.
+
+        // Zeroize master components
+        master.depth.zeroize();
+        {
+            let fingerprint: &mut [u8; 4] = master.parent_fingerprint.as_mut();
+            fingerprint.copy_from_slice(&[0u8; 4]);
```

Comment:
I would avoid this, as it could be optimized away. Instead, get a mutable slice and zeroize it. (Same elsewhere.)